samza-core/src/main/java/org/apache/samza/operators/impl/WindowOperatorImpl.java

```diff
@@ -101,29 +103,27 @@ public void onNext(M message, MessageCollector collector, TaskCoordinator coordi
       TriggerKey<WK> triggerKey = new TriggerKey<>(FiringType.EARLY, storeKey);
 
       getOrCreateTriggerImplWrapper(triggerKey, window.getEarlyTrigger())
-          .onMessage(message, collector, coordinator);
+          .onMessage(triggerKey, message, collector, coordinator);
     }
 
     if (window.getDefaultTrigger() != null) {
       TriggerKey<WK> triggerKey = new TriggerKey<>(FiringType.DEFAULT, storeKey);
       getOrCreateTriggerImplWrapper(triggerKey, window.getDefaultTrigger())
-          .onMessage(message, collector, coordinator);
+          .onMessage(triggerKey, message, collector, coordinator);
     }
   }
 
   @Override
   public void onTimer(MessageCollector collector, TaskCoordinator coordinator) {
-    long now = clock.currentTimeMillis();
-    TriggerCallbackState state;
-    while ((state = pendingCallbacks.peek()) != null && state.getScheduledTimeMs() <= now) {
-      pendingCallbacks.remove();
-      state.getCallback().run();
+    List<TriggerKey<WK>> keys = triggerScheduler.runPendingCallbacks();
```

Comment:
nit: I thought that the following pattern will allow window operator specific code lives here and let scheduler to invoke callbacks accordingly, something like:
triggerScheduler.runPendingCallbacks((key) -> {
  TriggerImplWrapper wrapper = triggers.get(key);
  if (wrapper != null) {
    wrapper.onTimer(key, collector, coordinator);
  }
})

Not too strong on this point. If you have specific reasons not to, I am fine with the current code.